### PR TITLE
Issue #4825 - Implement spec requirements for Request.newPushBuilder

### LINKED ISSUE #4825: PushBuilder tck test failures
**10.0.x**
serverPushCookieTest:
https://github.com/eclipse-ee4j/jakartaee-tck/blob/8.0.2/src/com/sun/ts/tests/servlet/spec/serverpush/TestServlet4.java
Fails because jetty does not copy Set-Cookie headers from the response to Cookie headers on the PushBuilder

serverPushInitTest:
https://github.com/eclipse-ee4j/jakartaee-tck/blob/8.0.2/src/com/sun/ts/tests/servlet/spec/serverpush/TestServlet2.java
Fails because jetty does not copy Authorization header from the request to the PushBuilder

serverPushNegativeTest:
https://github.com/eclipse-ee4j/jakartaee-tck/blob/8.0.2/src/com/sun/ts/tests/servlet/spec/serverpush/TestServlet7.java
Fails because:
> NullPointerException should be thrown if the argument of method() is null
> IllegalArgumentException should be thrown if set method "" to be used for the push
> IllegalArgumentException should be thrown if set method "POST" to be used for the push
> IllegalArgumentException should be thrown if set method "PUT" to be used for the push
> IllegalArgumentException should be thrown if set method "DELETE" to be used for the push
> IllegalArgumentException should be thrown if set method "CONNECT" to be used for the push
> IllegalArgumentException should be thrown if set method "OPTIONS" to be used for the push
> IllegalArgumentException should be thrown if set method "TRACE" to be used for the push

### PR DESCRIPTION
**Pull request by janbartel**
Closes #4825 

Implement spec for javax.servlet.http.PushBuilder according to the javadoc, and as tested for by the tck.

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/PushBuilderImpl.java [hunk 1]
```diff
@@ -39,6 +42,13 @@
     private static final Logger LOG = LoggerFactory.getLogger(PushBuilderImpl.class);
 
     private static final HttpField JettyPush = new HttpField("x-http2-push", "PushBuilder");
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/PushBuilderImpl.java]: Please make this field have the correct case (all upper case).

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 2]
```diff
@@ -410,12 +416,55 @@ public PushBuilder newPushBuilder()
...
+        //PushBuilder, unless the max-age of the cookie is <= 0
+        HttpFields responseFields = getResponse().getHttpFields();
+        for (HttpField field : responseFields)
+        {
+            HttpHeader header = field.getHeader();
+            if (header == null)
+                continue;
+            switch (header)
+            {
+                case SET_COOKIE:
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: You can easily coalesce the null-check and the switch into a single `if (header == SET_COOKIE)`.
How about `SET_COOKIE2` (I'm ok to not consider it, just raising the issue of whether we want to explicitly not consider it).

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 3]
```diff
@@ -410,12 +416,55 @@ public PushBuilder newPushBuilder()
...
+            StringBuilder buff = new StringBuilder();
+            for (Map.Entry<String,String> entry : cookies.entrySet())
+            {
+                if (buff.length() > 0)
+                    buff.append("; ");
+                buff.append(entry.getKey()).append('=').append(entry.getValue());
+            }
+            fields.add(new HttpField(HttpHeader.COOKIE, buff.toString()));
+        }
+
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:456]: I'm not sure handling these things at `PushBuilder` creation is the correct moment to do it.
Is the TCK testing for this?

I would rather add cookies, authorization headers, etc. just before the call to `push()`, to integrate also what the application is providing.

For example, if the application calls `pushBuilder.addHeader("cookie", "a=b")` is it added to the ones computed above, or it replaces them, etc.?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java (hunk 3 above)
Comment (janbartel) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:456]: But the tck test checks that the Cookie header is set on the PushBuilder before push() is called. See https://github.com/eclipse-ee4j/jakartaee-tck/blob/8.0.2/src/com/sun/ts/tests/servlet/spec/serverpush/TestServlet4.java#L48

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java (hunk 2 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: `Set-Cookie2` is an actual header from a valid spec.
https://tools.ietf.org/html/rfc2965

File: jetty-server/src/main/java/org/eclipse/jetty/server/Request.java (hunk 3 above)
Comment (sbordet) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:456]: > But the tck test checks that the Cookie header is set on the PushBuilder before push() is called.

That is so wrong.

Then the application calls `addHeader("Cookie", ...)` completely invalidating the fact that the `Cookie` header must be only one, so the server will complain that it found 2 Cookie headers.
Also, the application may mess with all other headers that you are trying to cope with here.

I think all this should be done just before `push()`.

I would challenge the TCK test.

Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:456]: I think servers are forgiving of multiple Cookie headers.  I know we are.
I think it is reasonable that a pushed request has the same cookies as the source request and some of them may be beyond the knowledge of the app doing the push.
The builder is only setting the initial headers and the app is free to change them as it likes, so I think starting with the known cookies is reasonable.

Comment (sbordet) [reply jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:456]: All right then.

**Review by sbordet — APPROVED** [review]

**Pull request merged**